Fix class_uses polyfill

WouterJ opened the pull request:
`class_uses` accepts both the class name and object instance as first argument, `class_exists` only accepts a class name. If an object instance is giving, the class is already autoloaded, so the other logic can be skipped.

nicolas-grekas commented:
Could you add a test please ?

WouterJ commented:
@nicolas-grekas done

WouterJ commented:
I've completely changed the function now btw, as it had to return an array and not a boolean as it did before.

nicolas-grekas commented:
Great! Somme comments:
- class_uses with a non-existent class name  returns false,
-  we miss a test with a class instance,
- I'm not sure the spl_autoload_register provides more behavior coverage, does it? If yes, then I'd suggest unregistering the autoloader to leave the global state clean after the test has run.

WouterJ commented:
I've fixed the first 2 comments.

The test with `spl_autoload_register` was there to test the second argument of class_uses: `$autoload`. However, I couldn't get things to work probably (unregistering didn't work for me for some reason), so I've removed the tests for now.

nicolas-grekas commented on src/Php54/bootstrap.php:
```diff
@@ -16,7 +16,14 @@
-        function class_uses($class, $autoload = true) { return $autoload && class_exists($class, $autoload) && false; }
+        function class_uses($class, $autoload = true)
+        {
+            if (!is_object($class) && $autoload && !class_exists($class, true)) {
```
when autoload === false, we should still return false when the class does not exist. But when $class is an interface, we shoudld return array()... Could be done with:
```php
if (is_object($class) || class_exists($class, $autoload) || interface_exists($class, false)) {
    return array();
}

return false;
```

nicolas-grekas commented on tests/Php54/Php54Test.php:
```diff
@@ -0,0 +1,32 @@
+
+    public function testClassUsesClassInstance()
+    {
+        $this->assertEquals(array(), class_uses(new Php54Test_Fixture()));
```
could be just `new stdClass()`

nicolas-grekas commented on tests/Php54/Php54Test.php:
```diff
@@ -0,0 +1,32 @@
+{
+    public function testClassUsesAutoloadedClass()
+    {
+        $this->assertEquals(array(), class_uses(__CLASS__.'_Fixture'));
```
assertSame

WouterJ commented:
Updated again

nicolas-grekas commented on src/Php54/bootstrap.php:
```diff
@@ -16,7 +16,14 @@
-        function class_uses($class, $autoload = true) { return $autoload && class_exists($class, $autoload) && false; }
+        function class_uses($class, $autoload = true)
+        {
+            if (is_object($class) || class_exists($class, $autoload) || interface_exists($class, $autoload)) {
```
interface_exists($class, **false**) because the autoloader has already been triggered, if needed, by class_exists

WouterJ replied:
fixed

stof commented on tests/Php54/Php54Test.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function testClassUsesInvalid()
+    {
+        $this->assertFalse(@class_uses('NotDefined'));
```
do we need the silencing here ?

nicolas-grekas replied:
yes: php triggers a warning

stof replied:
then shouldn't we try to polyfill it properly (and test that) ?

WouterJ replied:
How can you polyfill a warning trigger? And I was unsure how to test triggered errors in PHPunit, so I skipped that test.

nicolas-grekas replied:
@WouterJ by trigger an empty error, then using a custom error handler that aways returns false, the call `@class_uses`, then check the result of error_get_last.

nicolas-grekas commented:
:+1: (the missing warning is a bonus to me, nobody should rely on it)

xabbuh commented:
:+1:

WouterJ commented:
Can we please merge this?

fabpot commented:
Thank you @WouterJ.
